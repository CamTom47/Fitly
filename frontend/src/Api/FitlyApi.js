import axios from 'axios';

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

/** Fitly API Class
 * 
 * Static class typing together methods used to get/send to the Fitly API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't be any API-aware stuff elsehwere in the frontend
 */

class FitlyApi {
    //the token for interaction with API will be stored here.
    static token;

    static async request( endpoint, data = {}, method = "get"){
        console.debug("API Call:", endpoint, data, method)

        const url = `${BASE_URL}/${endpoint}`;
        let headers;
        
        (FitlyApi.token !== "") ? headers = { Authorization: `Bearer ${FitlyApi.token}` } : headers = null;
        const params = (method === "get")
            ? data
            : {};

            try{
                return (await axios({url, method, data, params, headers})).data;
            } catch(err) {
                console.error("API Error:", err.response);
                console.log(headers)
                let message = err.response.data.error.message;
                throw Array.isArray(message) ? message : [message];
            }
            }


            //Individual Fitly API routes

            //Front End user methods

            /**
             * Get a user's information
             * 
             * Parameter: {data}
             * 
             * {data: {token}}
             * 
             * Returns: 
             *  {"user": {
             *      "username",
             *      "firstName",
             *      "lastName",
             *      "password"
             *      "email",
        *           "isAdmin"}}
             */

            static async findUser(username){
                let res = await this.request(`users/${username}`)
                return res.user
            }

        /**
         * Authenticate a user and login
         * 
         * @param {*} data
         * 
         * {data: {username, password}}
         * 
         * @returns {token}
         */

        static async login(data){   
            let res = await this.request('auth/token', data, 'post')
            return res.token;
        }

        /** 
         * Sign up a user and create an account
         * 
         * @param {*} data
         *
         * {data: {username, password}}
         * 
         * @returns {token}
         */

        static async signup(data){
            let res = await this.request(`auth/register`, data, 'post')
            return res.token;
        }

        static async updateUser(username, data){
            let res = await this.request(`users/${username}`, data, 'patch');
            return res.user;
        }

        //Workout Methods
        
        /**
         * Find all workouts in the database that a user has created
         * 
         * 
         */

        static async findAllWorkouts(){
            let res = await this.request(`workouts/`);
            return res.workouts;

        }
        
        
        /**
         * 
         * @param {*} data 
         * {data : {user_id}}
         * 
         * @returns {workout}
         */

        static async findWorkout(data){
            console.log(data)
            let res = await this.request(`workouts/${data.workoutId}`);
            return res.workout;
        }
        
        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {workout}
         */

        static async createWorkout(data){
            let res = await this.request(`workouts/`, data, 'post');
            return res.newWorkout;
        }

        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {workout}
         */

        static async updateWorkout(workout_id,data){
            let res = await this.request(`workouts/${workout_id}`, data, 'patch');
            return res.updatedWorkout;
        }
        
        /**
         * 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */

        static async deleteWorkout(workoutId){
            let res = await this.request(`workouts/${workoutId}`, {}, 'delete');
            return res.workout;
        }

        static async addWorkoutCircuit(data){
            let res = await this.request(`workouts//${data.workoutId}/circuits/${data.circuitId}`, data, 'post');
            return res.workoutCircuit;
        }

        //Exercise Methods

        /**
         * Find all Exercises created by a user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {exercises}
         */

        static async findAllExercises(data){
            let res = await this.request(`exercises/`, data, 'get');
            return res.exercises
        }
        

        /**
         * Find a specfic exercise created by a user
         * @param {*} data 
         * {data {user_id}}
         * @returns {exercise}
         */

        static async findExercise(data){
            let res = await this.request(`exercises/${data.exercise_id}`, data, 'get');
            return res.exercise
        }
        
        /**
         * Creates a new exercise
         * @param {*} data 
         * {data: {user_id}}
         * @returns {exercise}
         */

        static async createExercise(data){
            let exercise = await this.request(`exercises/`, data, 'post');
            return exercise
        }
        
        /**
         * Update an existing exercise
         * @param {*} data 
         * {data: {exercise}}
         * @returns {exercise}
         */
        static async updateExercise(exerciseId, data){
            let res = await this.request(`exercises/${exerciseId}`, data, 'patch');
            return res.updatedExercise
        }
        
        /**
         * Delete an existing exercise 
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */

        static async deleteExercise(data){
            let res = await this.request(`exercises/${data.exercise_id}`, data, 'delete');
            return res.message
        }

        //Category Methods

        /**
         * Get all default categories and categories created by the user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {categories}
         */

        static async findAllCategories(data){
            let res = await this.request(`categories/`, data, 'get');
            return res.categories
        }

        /**
         * Get a category by category_id
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */
        static async findCategory(data){
            let res = await this.request(`categories/${data.category_id}`, data, 'get');
            return res.category;
        }

        /**
         * Create a new category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */

        static async createCategory(data){
            let res = await this.request(`categories/`, data, 'post');
            return res.category;
        }

        /**
         * Update an existing workout category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {category}
         */

        static async updateCategory(data){
            let res = await this.request(`categories/:category_id`, data, 'patch');
            return res.category;
        }

        /**
         * Delete an existing workout category
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */
        static async deleteCategory(data){
            let res = await this.request(`categories/:category_id`, data, 'delete');
            return res.category;
        }

        //Equipment Methods

        /**
         * Get all default equipments and equipment created by the user
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipments}
         */

        static async findAllEquipments(data){
            let res = await this.request(`equipments/`, data, 'get');
            return res.equipments
        }

        /**
         * Get a category by category_id
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */
        static async findEquipment(data){
            let res = await this.request(`equipments/${data.equipment_id}`, data, 'get');
            return res.equipment;
        }

        /**
         * Create a new equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */

        static async createEquipment(data){
            let res = await this.request(`equipments/`, data, 'post');
            return res.equipment;
        }

        /**
         * Update an existing workout equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {equipment}
         */

        static async updatEquipment(data){
            let res = await this.request(`equipments/:equipment_id`, data, 'patch');
            return res.equipment;
        }

        /**
         * Delete an existing workout equipment
         * @param {*} data 
         * {data: {user_id}}
         * @returns {message}
         */
        static async deleteEquipment(data){
            let res = await this.request(`equipments/:equipment_id`, data, 'delete');
            return res.equipment;
        }

        //Circuit Methods


        static async findAllCircuits(data){
            let res = await this.request(`circuits/`, data, 'get');
            return res.circuits;
        }

        static async findCircuit(data){
            let res = await this.request(`circuits/${data.circuit_id}`, data, 'get');
            return res.circuit;
        }

        static async addCircuit(data){
            let res = await this.request(`circuits/`, data, 'post');
            return res.circuit;
        }
        
        static async updateCircuit(circuitId,data){
            let res = await this.request(`circuits/${circuitId}`, data, 'patch');
            return res.updatedCircuit;
        }

        static async deleteCircuit(circuitId){
            let res = await this.request(`circuits/${circuitId}`, {} , 'delete');
            return res.message;
        }

        static async addExerciseCircuit(data){
            let res = await this.request(`circuits/${data.circuitId}/exercises/${data.exerciseId}`, data, 'post');
            return res.circuitExercise;
        }
        
        static async updateExerciseCircuit(data){
            let res = await this.request(`circuits/${data.circuitId}/exercises/${data.exerciseId}`, data, 'patch');
            return res.circuitExercise;
        }

        static async findAllMuscleGroups(data){
            let res = await this.request(`muscleGroups/`, data, 'get');
            return res.muscleGroups
        }
        
        static async findMuscleGroup(data){
            let res = await this.request(`muscleGroups/${data.muscleGroupId}`, data, 'get');
            return res.muscleGroup
        }
    }

    FitlyApi.token = ""

export default FitlyApi;