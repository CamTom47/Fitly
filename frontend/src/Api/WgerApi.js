import axios from 'axios';

const WGER_BASE_URL = "https://wger.de/api/v2"

    class WgerApi {
        static async request( endpoint, data = {}, method = "get"){
            console.debug("WGER Call:", endpoint, data, method)

            const url = `${WGER_BASE_URL}/${endpoint}`

            const params = (method === "get")
            ? data
            : {};
            try{
                const res = (await axios({url, method, data, params}));
                return res
            } catch(err){
                console.error("wger API Error:", err.response)
                let message = err.response.data.error.message
                throw Array.isArray(message) ? message : [message];
            }
        }

        static async getAllExercises(endpoint = 'exercisebaseinfo/'){
            if(endpoint.includes(WGER_BASE_URL)){
                endpoint = endpoint.split(`${WGER_BASE_URL}/`)[1]
            };
            
            console.log(endpoint)
            let res = await this.request(`${endpoint}`)
            console.log(res)
            return res
        }

    }


    export default WgerApi