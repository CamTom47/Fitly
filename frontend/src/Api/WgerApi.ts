import axios, { AxiosResponse } from 'axios';

const WGER_BASE_URL : string = "https://wger.de/api/v2"

    class WgerApi {
        static async request( endpoint: string, data: object = {}, method: string = "get"){
            console.debug("WGER Call:", endpoint, data, method)

            const url: string = `${WGER_BASE_URL}/${endpoint}`

            const params: {} = (method === "get")
            ? data
            : {};
            try{
                const res = (await axios({url, method, data, params}));
                return res
            } catch(err){
                console.error("wger API Error:", err.response)
                let message : string = err.response.data.error.message
                throw Array.isArray(message) ? message : [message];
            }
        }

        static async getAllExercises<Promise>(endpoint : string = 'exercisebaseinfo/'): Promise {

            // The initial request will set the next and previous calls with a end point similar to 
            // "next": "https://wger.de/api/v2/exercisebaseinfo/?limit=20&offset=20",
            // Splitting on the WGER_BASE_URL allows for the new and point to be added to the default WGER_BASE_URL

            if(endpoint.includes(WGER_BASE_URL)){
                endpoint = endpoint.split(`${WGER_BASE_URL}/`)[1]
            };
            
            let res = await this.request(endpoint)

            return res.data
        }

    }


    export default WgerApi