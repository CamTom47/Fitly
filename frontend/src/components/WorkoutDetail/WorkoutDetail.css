#WorkoutDetailContainer {
	background: var(--color-primary);
	padding: 0.5rem 0 3rem 0;
	border-radius: 1rem;
	border: 1px solid var(--color-primary);
	box-shadow: 0 1px 7px;
	margin: 1rem 4rem 0 4rem;
	padding: 0;
	height: 6rem;
	display: flex;
	align-items: center;
	justify-content: space-around ;
}
.WorkoutHeader-1{
	display: flex;
	align-items: baseline;
}
.WorkoutHeader-2 {
	display: flex;
}
.WorkoutHeader-2 div {
	width: 15rem;
	display: flex;
	flex-direction: column;
	align-items: center;
}

.WorkoutHeader-2 span {
	text-decoration: underline;
	font-size: 14px;
	text-align: right;
	padding-right: 0.4rem;
}
.WorkoutHeader-2 p {
	text-decoration: none;
	font-size: 14px;
}
.circuitSection h3 {
	color: var(--black);
}
.WorkoutDetailButton {
	background-color: var(--color-primary);
	color: var(--color-secondary);
	padding: 0 10px;
	margin-left: 1rem;
	border: 1px solid var(--color-secondary);
	border-radius: 12px;
}
.WorkoutDetailButton:hover {
	background-color: var(--color-secondary);
	color: var(--color-primary);
	padding: 0 10px;
	margin-left: 1rem;
	border: 1px solid var(--color-pri);
	border-radius: 12px;
	transition: all 0.35s;
}
.WorkoutDetailButton > .LinkElement {
	color: var(--color-secondary);
	text-decoration: none;
}
.WorkoutDetailButton > .LinkElement:hover {
	color: var(--color-primary);
	text-decoration: none;
	transition: all 0.35s;
}

/* Styling For table */

.ActiveWorkout-Container {
	display: flex;
	flex-direction: column;
	justify-content: center;
	height: 30rem;
	width: 50%;
	border: 1px solid var(--color-secondary);
	border-radius: 5px;
	background: var(--color-primary);
	margin-left: 0.5rem;
	box-shadow: 0 0 5px black;
}


.ActiveWorkout-WorkoutStarter {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	margin: 0 auto;
	background: var(--color-primary);
	border: 1px solid var(--gray-4);
	border-radius: 50%;
	padding: 2rem 0 2rem 0;
	width: 10rem;
	height: 10rem;
}

.WorkoutTable {
	width: 100%;
	margin-top: 1rem;
}
tr:hover {
	background: var(--color-secondary);
	color: white;
	transition: all 0.35s;
}
th,
td {
	border-bottom: 1px solid #ddd;
	text-align: center;
	padding: 0.5rem 0 0.5rem 0.5rem;
	width: 10%;
	font-size: 12px;
}
tr:focus {
	background-color: black;
}

th {
	background: var(--gray-4);
	color: var(--white);
}
h5 {
	padding: 1rem 0 0.5rem 1rem;
}
.circuitSectionHeader {
	display: flex;
	align-items: baseline;
	margin-top: 1rem;
}
.tableContainer {
	height: 25rem;
	width: 100%;
	overflow-y: scroll;
}
.WorkoutDetailContainerInner {
	display: flex;
}
.WorkoutDetailContainerHead {
	display: flex;
	align-items: center;
}
.WorkoutHeader {
	display: flex;
	align-items: baseline;
}
.star {
	color: var(--gray-2);
}
.star {
	color: var(--color-secondary);
}
.favoritedStar {
	color: var(--color-secondary);
}
.favoritedStar:hover {
	color: var(--gray-2);
}
.LowerSection {
	display: flex;
	justify-content: space-between;
	margin: 1rem 4rem;
}
.circuitSection {
	width: 50%;
	margin-right: 0.5rem;
	border: 1px solid var(--color-secondary);
	border-radius: 5px;
	background: var(--color-primary);
	box-shadow: 0 0 5px black;
}
.ActiveWorkout {
	position: relative;
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
}
.ActiveWorkout-buttonContainer {
	position: absolute;
	top: 88%;
	display: flex;
}
.ActiveWorkout-button {
	background: var(--color-primary);
	color: var(--color-secondary);
	border: 1px solid var(--color-secondary);
	padding: 0.5rem;
	border-radius: 5px;
}
.ActiveWorkout-button:hover {
	color: var(--color-primary);
	background-color: var(--color-secondary);
	text-decoration: none;
	transition: all 0.35s;
}
.playButton {
	color: var(--color-secondary);
	padding: 0.5rem;
	transform: translateX(.1rem);
}
.iconButtonOuterRing {
	display: flex;
	justify-content: center;
	align-items: center;
	height: 50px;
	width: 50px;
	border: 1px solid var(--color-secondary);
	border-radius: 50%;
	color: var(--color-secondary);
}
.iconButtonOuterRing:hover {
	background-color: var(--color-secondary);
	text-decoration: none;
	transition: all 0.35s;
}
.iconButtonOuterRing:hover > * {
	color: var(--color-primary);
	transition: all 0.35s;
}
.none {
	display: none;
}
.disabled{
	pointer-events: none;
}