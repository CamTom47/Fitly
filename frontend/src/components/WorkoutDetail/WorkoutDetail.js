//functionality imports
import React, { useCallback, useEffect, useState, useContext } from "react";
import { useNavigate, Link, useFetcher } from "react-router-dom";
import UserContext from "../../context/UserContext";

//styling imports
import { Card, CardBody, CardText, CardTitle, ListGroup, ListGroupItem, Button, Container} from "reactstrap"
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faStar, faTrash, faSquarePlus } from '@fortawesome/free-solid-svg-icons'
import LoadingComponent from "../LoadingComponent/LoadingComponent";

//component imports
import NewCircuitForm from "../Forms/NewCircuitForm/NewCircuitForm";
import UpdateWorkoutForm from "../Forms/UpdateWorkoutForm/UpdateWorkoutForm"
import Circuit from "../../components/Circuit/Circuit";
import FitlyApi from "../../Api/FitlyApi";




const WorkoutDetail = () => {

    let navigate = useNavigate();
    let currentUser = useContext(UserContext).currentUser;

    const [workout, setWorkout] = useState({});
    const [circuits, setCircuits] = useState([]);
    const [showNewCircuitForm, setShowNewCircuitForm] = useState(false);
    const [showWorkoutUpdateForm, setShowWorkoutUpdateForm] = useState(false);
    const [category, setCategory] = useState({});
    const [isLoading, setIsLoading] = useState(true);

    //parse the url to get the workout parameter that will be used to find the workout using the Fitly API
    //...workouts/2 => 2

    let urlArray = (((window.location.pathname).split('/')));
    let workoutId = parseInt(urlArray[urlArray.length-1])

    const getWorkoutAndCircuits = useCallback( async () => { 
        try{
            let workout = await FitlyApi.findWorkout({
                workout_id: workoutId
            });
            setWorkout(workout);

            let circuits = await FitlyApi.findAllCircuits({
                workout_id: workout.id});
                let filteredCircuits = circuits.filter( circuit => circuit.workout_id === workout.id)
                setCircuits(filteredCircuits)

                
            } catch (err){
                return err
            }
    }, [showNewCircuitForm, showWorkoutUpdateForm, isLoading]);
    
    useEffect( () => {
        setIsLoading(true);
        getWorkoutAndCircuits();
        findCategory();
        setIsLoading(false);
    }, [getWorkoutAndCircuits]);
    
    
        
    
    //function that updates the workouts in the data base. formData is passed as a arugment from the newCircuitForm.
    const updateWorkout = async (formData) => {
        try{
            await FitlyApi.updateWorkout(workout.id, formData)
            getWorkoutAndCircuits();
        } catch (err) {
            return err
        }
    }

   //remove a workout for the fitly database
   const removeWorkout = async () => {
        try{
            await FitlyApi.deleteWorkout(workout.id);
            navigate('/workouts')
        } catch(err){
            return err
        }
    }


    // Add/remove workout from favorites to allow for filtering 
    const handleFavorite = async () => {
        try{
            setIsLoading(true);
            (workout.favorited === false)
                ? await FitlyApi.updateWorkout(workout.id, {favorited: true})
                : await FitlyApi.updateWorkout(workout.id, {favorited: false})
                setIsLoading(false);
        } catch(err){
            return err
        }
    }

    const toggleShowWorkoutUpdateForm = () => {
        setShowWorkoutUpdateForm(showWorkoutUpdateForm => !showWorkoutUpdateForm)
    }

    const toggleShowNewCircuitForm = () => {
        setShowNewCircuitForm(showNewCircuitForm => !showNewCircuitForm);
    }

    //Circuit Methods

    const createCircuit = async (formData, exerciseId) => {

        //create new circuit
        const circuit = await FitlyApi.addCircuit(formData);
         
         // add exercise to new circuit
         await FitlyApi.addExerciseCircuit({
             circuit_id: circuit.id, 
             exercise_id: exerciseId
         });

         // //add circuit to workout
         await FitlyApi.addWorkoutCircuit({
             workout_id: workout.id,
             circuit_id: circuit.id
         });

         getWorkoutAndCircuits();
     }

    const updateCircuit = 
        async (formData) => {
            try{
                await FitlyApi.updateCircuit(formData);
                getWorkoutAndCircuits();
            } catch (err){
                return err
            }
        }

    const deleteCircuit = async (formData) => {
        try{
            await FitlyApi.deleteCircuit(formData);
            getWorkoutAndCircuits();
        } catch (err) {
            return err
        }
    }

    const findCategory = useCallback(async () => {
        try{
            let category = await FitlyApi.findCategory({category_id: workout.category});
            setCategory(category)
        } catch(err){
            return err
        }
}, [workout])

useEffect( () => {
    setIsLoading(true);
    findCategory();
    setIsLoading(false);
}, [findCategory]);


    const circuitComponents = circuits.map( circuit => (
        <div>
            <Circuit circuitId={circuit.id} updateCircuit={updateCircuit} deleteCircuit={deleteCircuit}/>
            <hr></hr>
        </div>
    ))

    return (isLoading)
    ? <LoadingComponent/>
    : ( 
    (showWorkoutUpdateForm) 
    ? <UpdateWorkoutForm workout={workout} updateWorkout={updateWorkout} handleToggle={toggleShowWorkoutUpdateForm}/>
    :(
        <Container className="w-75 pt-3">
        <div>
            <div className="d-flex column-gap-3">
                <Button color='danger'>
                    <Link to={`/workouts`} style={{textDecoration: "none", color: "white"}}>
                        Back to Workouts
                    </Link>
                </Button>
                <Button onClick={toggleShowWorkoutUpdateForm}>
                    Edit Workout Information
                </Button>
            </div>

        <div>        
        <Card className="my-2 p-2 w-25">
            <div className="d-flex flex-row">
                <button className="btn btn-danger" onClick={removeWorkout}>Delete Workout</button>
                {
                    (workout.favorited)
                    ? <FontAwesomeIcon type="button" onClick={handleFavorite} icon={faStar} size="lg" style={{color: "#FFD43B"}}/>
                    : <FontAwesomeIcon type="button" onClick={handleFavorite} icon={faStar} size="lg"/>
                }
            </div>   
        <CardTitle className="fs-3" >{workout.name}</CardTitle>  
        <CardBody>
            <CardText>
                Type of Workout:{category.name}
            </CardText>          
        </CardBody>
    </Card>
    </div>
    <Card>
        <CardBody className="d-flex flex-grow-1 flex-column align-item-center row-gap-3">

            <CardText className="d-flex flex-column align-items-center fs-4">
                    Circuits
                    <hr className="w-100 mt-0"></hr>
            </CardText>
            {circuitComponents}
            <div className="d-flex flex-row justify-content-center align-items-center column-gap-3">
                <span>Add A Circuit</span>
                <FontAwesomeIcon type="button" icon={faSquarePlus} onClick={toggleShowNewCircuitForm}></FontAwesomeIcon>
            </div>
            { 
                (showNewCircuitForm)
                ? <NewCircuitForm workout={workout} createCircuit={createCircuit} toggleShowNewCircuitForm={toggleShowNewCircuitForm}/>
                : null
                }
            
        </CardBody>
    </Card>
        </div>
    </Container>
    )
)

}

export default WorkoutDetail;