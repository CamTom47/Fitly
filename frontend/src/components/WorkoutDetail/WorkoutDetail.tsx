//functionality imports
import React, { useCallback, useEffect, useState, useContext } from "react";
import { useNavigate, Link, useParams } from "react-router-dom";
import { v4 as uuid } from 'uuid'

//styling imports
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faStar, faTrash, faSquarePlus } from '@fortawesome/free-solid-svg-icons'
import LoadingComponent from "../LoadingComponent/LoadingComponent";
import './WorkoutDetail.css'

//component imports
import NewCircuitForm from "../Forms/NewCircuitForm/NewCircuitForm";
import UpdateWorkoutForm from "../Forms/UpdateWorkoutForm/UpdateWorkoutForm"
import Circuit from "../Circuit/Circuit";
import FitlyApi from "../../Api/FitlyApi";

import { useAppDispatch, useAppSelector } from "../../hooks/reduxHooks";
import { deleteWorkout, selectWorkouts } from '../../slices/workoutsSlice';
import { selectCircuits, findAllCircuits } from '../../slices/circuitsSlice';
import { selectCategories } from '../../slices/categoriesSlice';


interface Workout{
    id: number,
    user_id: number,
    name: string,
    category: number,
    favorited: boolean,
}

interface Category{
    id: number,
    name: string
}

interface CircuitType {
    id: number,
    sets: number,
    reps: number,
    weight: number
    rest_period: number,
    intensity: string,
    exerciseId? : number,
    workoutId? : number
}

const WorkoutDetail = (): React.JSX.Element => {
    const dispatch = useAppDispatch();
    let navigate = useNavigate();
    const workoutId = useParams().workout_id;
    const workouts = useAppSelector(selectWorkouts);
    const workout = workouts.find( (workout: Workout) => workout.id === Number(workoutId));
    const circuits = useAppSelector(selectCircuits).filter( (circuit: CircuitType) => circuit.workoutId === workout.id);
    const categories = useAppSelector(selectCategories);
    const category = categories.find( (category: Category) => category.id === workout.category);
    const [showNewCircuitForm, setShowNewCircuitForm] = useState(false);
    const [showWorkoutUpdateForm, setShowWorkoutUpdateForm] = useState(false);
    const [isLoading, setIsLoading] = useState(false);

    //remove a workout for the fitly database
   const removeWorkout = () => {
        dispatch(deleteWorkout(Number(workoutId)))    
        navigate('/workouts')
    }

    // Add/remove workout from favorites to allow for filtering 
    const handleFavorite = async () => {
        try{
            setIsLoading(true);
            (workout.favorited === false)
                ? await FitlyApi.updateWorkout(workout.id, {favorited: true})
                : await FitlyApi.updateWorkout(workout.id, {favorited: false})
                setIsLoading(false);
        } catch(err){
            return err
        }
    }

    const toggleShowWorkoutUpdateForm = () : void =>{
        setShowWorkoutUpdateForm(showWorkoutUpdateForm => !showWorkoutUpdateForm)
    }

    const toggleShowNewCircuitForm = () : void => {
        setShowNewCircuitForm(showNewCircuitForm => !showNewCircuitForm);
    }

    const circuitComponents = circuits.map( circuit => (
            <Circuit key={uuid()} circuitId={circuit.id}/>
    ))

    return (isLoading)
    ? <LoadingComponent/>
    : ( 
    (showWorkoutUpdateForm) 
    ? <UpdateWorkoutForm key={uuid()} workout={workout} handleToggle={toggleShowWorkoutUpdateForm}/>
    :(
        <div  id="WorkoutDetailContainer">
            <div className="WorkoutDetailContainerInner">
                <div className="WorkoutDetailContainerHead">
                    <button>
                        <Link className="LinkElement" to={`/workouts`}>Back to Workouts</Link>
                    </button>
                    <div>
                        <button onClick={toggleShowWorkoutUpdateForm}>Edit Workout Information</button>
                        <button onClick={(removeWorkout)}>Delete Workout</button>
                    </div>
                </div>
                <div className="WorkoutDetailDivider">
                    <div className="dividerSection1">
                        { (workout.favorited)
                            ? <FontAwesomeIcon type="button" onClick={handleFavorite} icon={faStar} size="lg" style={{color: "#FFD43B"}}/>
                            : <FontAwesomeIcon type="button" onClick={handleFavorite} icon={faStar} size="lg"/>
                        }
                        <h3>{workout.name}</h3>  
                        <div className="dividerSection2">
                            <span>Type of Workout:</span>
                            <p>{category.name}</p>          
                        </div>
                    </div>

                </div>
                <div className="circuitSection">
                    <div className="dividerSection1">
                        <h3>Circuits</h3>
                        <FontAwesomeIcon className="faIcon" size="xl" type="button" icon={faSquarePlus} onClick={toggleShowNewCircuitForm}></FontAwesomeIcon>
                    </div>
                    { (showNewCircuitForm)
                    ? <NewCircuitForm workout={workout} toggleShowNewCircuitForm={toggleShowNewCircuitForm}/>
                    : null
                    }
                    <div className="CircuitList">
                        {circuitComponents}
                    </div>
                
                </div>
                
        </div>
    </div >
    )
)

}

export default WorkoutDetail;