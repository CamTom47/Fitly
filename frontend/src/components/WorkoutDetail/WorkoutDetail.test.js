import { render, waitFor} from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
// import userEvent from '@testing-library/user-event'
import FitlyApi from '../../Api/FitlyApi';
import WgerApi from '../../Api/WgerApi';
import UserProvider from '../../../testUtils';
import WorkoutDetail from './WorkoutDetail';
 
//Mock FitlyAPI
jest.mock('../../Api/FitlyApi', () => ({
    findWorkout: jest.fn(() => (
            {
             id: 1,
            user_id: 1,
            name: "test workout",
            category: 1,
            favorited: false
        }
        ))
}));

//Mock WgerApi
jest.mock('../../Api/WgerApi', () => {
});



describe('WorkoutDetail Functionality', () => {

    let workout;

    waitFor(async ()=> {
        workout = await FitlyApi.findWorkout();
    })

    let children = (
        <WorkoutDetail workout={workout}/>
    );
    


  it('it renders without crashing', async () => {

    render(
        <UserProvider children={children}/>
      );
    }); 
    
    
    it('matches snapshot', async () => {
        const {asFragment} = render(
    <UserProvider children={children}/>
)
     expect(asFragment()).toMatchSnapshot();
    })
});