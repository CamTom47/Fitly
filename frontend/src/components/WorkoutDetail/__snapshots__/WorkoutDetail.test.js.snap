// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutDetail Functionality matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-75 pt-3 container"
  >
    <div>
      <div
        class="d-flex column-gap-3"
      >
        <button
          class="btn btn-danger"
        >
          <a
            href="/workouts"
            style="text-decoration: none; color: white;"
          >
            Back to Workouts
          </a>
        </button>
        <button
          class="btn btn-secondary"
          type="button"
        >
          Edit Workout Information
        </button>
      </div>
      <div>
        <div
          class="my-2 p-2 w-25 card"
        >
          <div
            class="d-flex flex-row"
          >
            <button
              class="btn btn-danger"
            >
              Delete Workout
            </button>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-star fa-lg "
              data-icon="star"
              data-prefix="fas"
              focusable="false"
              role="img"
              type="button"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            class="fs-3 card-title"
          />
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              Type of Workout:
            </p>
          </div>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="d-flex flex-grow-1 flex-column align-item-center row-gap-3 card-body"
        >
          <p
            class="d-flex flex-column align-items-center fs-4 card-text"
          >
            Circuits
          </p>
          <hr
            class="w-100 mt-0"
          />
          <p />
          <div
            class="d-flex flex-row justify-content-center align-items-center column-gap-3"
          >
            <span>
              Add A Circuit
            </span>
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-square-plus "
              data-icon="square-plus"
              data-prefix="fas"
              focusable="false"
              role="img"
              type="button"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M64 32C28.7 32 0 60.7 0 96L0 416c0 35.3 28.7 64 64 64l320 0c35.3 0 64-28.7 64-64l0-320c0-35.3-28.7-64-64-64L64 32zM200 344l0-64-64 0c-13.3 0-24-10.7-24-24s10.7-24 24-24l64 0 0-64c0-13.3 10.7-24 24-24s24 10.7 24 24l0 64 64 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-64 0 0 64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
